build/result.go

```diff
@@ -196,10 +202,115 @@ func (r *ResultContext) build(buildFunc gateway.BuildFunc) (err error) {
 	return err
 }
 
+func (r *ResultContext) solve(ctx context.Context, def *result.Result[*pb.Definition]) (*gateway.Result, error) {
+	resultCh := make(chan *gateway.Result)
+	errCh := make(chan error)
+	go func() {
+		err := r.build(func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// force evaluation of all targets in parallel
+			results := make(map[*pb.Definition]*gateway.Result)
+			resultsMu := sync.Mutex{}
+			eg, egCtx := errgroup.WithContext(ctx)
+			def.EachRef(func(def *pb.Definition) error {
+				eg.Go(func() error {
+					res2, err := c.Solve(egCtx, gateway.SolveRequest{
+						Evaluate:   true,
+						Definition: def,
+					})
+					if err != nil {
+						return err
+					}
+					resultsMu.Lock()
+					results[def] = res2
+					resultsMu.Unlock()
+					return nil
+				})
+				return nil
+			})
+			if err := eg.Wait(); err != nil {
+				return nil, err
+			}
+			res2, _ := result.ConvertResult(def, func(def *pb.Definition) (gateway.Reference, error) {
+				if res, ok := results[def]; ok {
+					return res.Ref, nil
+				}
+				return nil, nil
+			})
+			select {
+			case resultCh <- res2:
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			}
+			<-ctx.Done()
+			return nil, nil
+		})
+		if err != nil {
+			errCh <- err
+		}
+	}()
+	select {
+	case req := <-resultCh:
+		return req, nil
+	case err := <-errCh:
+		return nil, err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (r *ResultContext) getParentResultOfError(ctx context.Context) (*gateway.Result, error) {
+	if r.def == nil || r.rawErr == nil || r.solveErr == nil {
+		return nil, errors.Errorf("no definition is provided")
+	}
+	var dgst digest.Digest
+	var ve *errdefs.VertexError
+	if errors.As(r.rawErr, &ve) {
+		dgst = digest.Digest(ve.Digest)
+	} else {
+		return nil, errors.Errorf("unsupported vertex: cannot get parent of error")
+	}
+
+	var parentSelector func(inputs []llb.Output) llb.Output
+	switch r.solveErr.Solve.Op.GetOp().(type) {
+	case *pb.Op_File:
+		parentSelector = func(inputs []llb.Output) llb.Output { return inputs[0] } // TODO: allow user selecting one input?
```

Comment:
An op can have multiple inputs and the current implementation always uses the first input (`input[0]`). But maybe we need to introduce a way to ensure the input is actually the previous step's rootfs. Or maybe we can introduce a way to let the user choose one of the inputs. 
Currently I don't have good idea for them so please let me know if there is a good way to solve it.